### PR TITLE
[Upgrade] Go-Ethereum release v1.10.1

### PR DESCRIPTION
**Pull request by quorumbot**
### Plan & Analyse

- [x] Review the Release Notes
- [x] Review PRs in the section below

As you review, list extra changes and/or tests to be implemented to ensure compatibility with GoQuorum specific features.

### Build & Test

- [x] Pull and checkout PR branch locally, then merge GoQuorum `master` into this branch
- [x] Resolve conflicts, taking into account the prior analysis
- [x] Implement required changes until lint passes
- [x] Implement required changes until all unit tests pass
- [x] Implement required changes until acceptance tests pass
- [x] Implement extra changes and/or tests
- [x] Verify any left TODOs in the code

Add any extra changes/tests as comments on this PR.

## Go-Ethereum Release: Gemini Sigma (v1.10.1)

* Version: v1.10.1
* Published: 2021-03-08T09:35:47Z

### Release notes 

Geth v1.10.1 is a minor release with the sole purpose of enabling the [***Berlin hard-fork***](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/berlin.md)! This hard-fork takes a step towards making opcodes fairer and lays the groundwork to new types of transactions, with lots of interesting features to be built on top.

The Ethereum Foundation will have a dedicated blog post for Berlin. The essential parts from Geth's perspective is that v1.10.1 is ***required*** for Berlin on all testnets and the mainnet too. Below you can find the fork blocks for the different networks and their expected schedules. Please ensure you are upgraded well in advance of the forks to ensure a smooth transition.	

- ***Ropsten*** 9,812,189 (10 Mar 2021)	
- ***Goerli*** 4,460,644 (17 Mar 2021)	
- ***Rinkeby*** 8,290,928 (24 Mar 2021)	
- ***Mainnet*** 12,244,000 (14 Apr 2021)	

For a full rundown of the changes please consult the Geth 1.10.1 [release milestone](https://github.com/ethereum/go-ethereum/milestone/113?closed=1).

---

As with all our previous releases, you can find the:

- Pre-built binaries for all platforms on our [downloads page](https://geth.ethereum.org/downloads/).
- Docker images published under [`ethereum/client-go`](https://cloud.docker.com/u/ethereum/repository/docker/ethereum/client-go).
- Ubuntu packages in our [Launchpad PPA repository](https://launchpad.net/~ethereum/+archive/ubuntu/ethereum).
- OSX packages in our [Homebrew Tap repository](https://github.com/ethereum/homebrew-ethereum).

### Legend

File Stats: (A) Added, (M) Modified and (R) Removed

Line Stats: (A) Added and (R) Removed

Assessment:

* ✅ No conflict expected
* ⚠ Review required to assess changes
* ‼️ Conflicts expected and review required

### 5 Pull Requests

| 🔍 | Link | Title | File Stats<br>M/A/R | Packages changed<br>(files changed) | Line Stats<br>A/R | Top 5 Changed Files<br>(lines changed) |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| ⚠️ | [#22431](https://github.com/ethereum/go-ethereum/pull/22431) | ``build: fix PPA failure due to updated debsrc`` | 1/0/0<br> | ``build`` (1)<br> | <span class="text-green">6</span>/<span class="text-red">5</span><br> | ``build/ci.go`` (22)<br> |
| ⚠️ | [#22432](https://github.com/ethereum/go-ethereum/pull/22432) | ``build: add support for Ubuntu Hirsute Hippo`` | 1/0/0<br> | ``build`` (1)<br> | <span class="text-green">6</span>/<span class="text-red">6</span><br> | ``build/ci.go`` (24)<br> |
| ‼️ | [#22414](https://github.com/ethereum/go-ethereum/pull/22414) | ``core, eth: unship EIP 2315`` | 15/0/0<br> | ``core/vm`` (12)<br>``eth/tracers`` (2)<br>``core/vm/runtime`` (1)<br> | <span class="text-green">50</span>/<span class="text-red">428</span><br> | ``core/vm/runtime/runtime_test.go`` (450)<br>``core/vm/logger.go`` (70)<br>``core/vm/opcodes.go`` (68)<br>``core/vm/instructions_test.go`` (66)<br>``core/vm/instructions.go`` (64)<br> |
| ‼️ | [#22426](https://github.com/ethereum/go-ethereum/pull/22426) | ``Revert "core/forkid, params: unset Berlin fork number"`` | 2/0/0<br> | ``core/forkid`` (1)<br>``params`` (1)<br> | <span class="text-green">37</span>/<span class="text-red">25</span><br> | ``core/forkid/forkid_test.go`` (116)<br>``params/config.go`` (8)<br> |
| ⚠️ | [#22452](https://github.com/ethereum/go-ethereum/pull/22452) | ``tests: update reference tests with 2315 removed from Berlin`` | 1/0/0<br> | ``tests`` (1)<br> | <span class="text-green">1</span>/<span class="text-red">1</span><br> | ``tests/testdata`` (4)<br> |

### 20 Changed files

| 🔍 | File | Lines Changed | Linked PR |
| :--- | :--- | :--- | :--- |
| ‼️ | ``core/vm/runtime/runtime_test.go`` | 450 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ‼️ | ``core/forkid/forkid_test.go`` | 116 | [#22426](https://github.com/ethereum/go-ethereum/pull/22426)<br> |
| ✅ | ``core/vm/logger.go`` | 70 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``core/vm/opcodes.go`` | 68 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ‼️ | ``core/vm/instructions_test.go`` | 66 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ‼️ | ``core/vm/instructions.go`` | 64 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``core/vm/stack.go`` | 62 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``core/vm/eips.go`` | 58 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ⚠️ | ``build/ci.go`` | 46 | [#22431](https://github.com/ethereum/go-ethereum/pull/22431)<br>[#22432](https://github.com/ethereum/go-ethereum/pull/22432)<br> |
| ⚠️ | ``core/vm/interpreter.go`` | 32 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``core/vm/gen_structlog.go`` | 26 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``core/vm/contract.go`` | 26 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``core/vm/logger_json.go`` | 10 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ✅ | ``eth/tracers/tracer.go`` | 8 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ‼️ | ``eth/tracers/tracer_test.go`` | 8 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ‼️ | ``params/config.go`` | 8 | [#22426](https://github.com/ethereum/go-ethereum/pull/22426)<br> |
| ‼️ | ``core/vm/logger_test.go`` | 6 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |
| ‼️ | ``params/version.go`` | 4 |  |
| ⚠️ | ``tests/testdata`` | 4 | [#22452](https://github.com/ethereum/go-ethereum/pull/22452)<br> |
| ✅ | ``core/vm/jump_table.go`` | 2 | [#22414](https://github.com/ethereum/go-ethereum/pull/22414)<br> |

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1639,6 +1639,11 @@ func (s *PublicTransactionPoolAPI) GetTransactionReceipt(ctx context.Context, ha
 	}
 	receipt := receipts[index]
 
+	// Derive the sender.
+	//bigblock := new(big.Int).SetUint64(blockNumber)
+	//signer := types.MakeSigner(s.b.ChainConfig(), bigblock)
+	//from, _ := types.Sender(signer, tx)
+
```
Comment (baptiste-b-pegasys) [inline internal/ethapi/api.go]:
```suggestion
```

File: eth/handler.go [hunk 2]
```diff
@@ -575,6 +575,11 @@ func (h *handler) BroadcastTransactions(txs types.Transactions) {
 		directCount += len(hashes)
 		peer.AsyncSendTransactions(hashes)
 	}
+	for peer, hashes := range txset {
+		directPeers++
+		directCount += len(hashes)
+		peer.AsyncSendTransactions(hashes)
+	}
```
Comment (baptiste-b-pegasys) [inline eth/handler.go]:
```suggestion
```
duplicate block

File: core/tx_pool.go [hunk 3]
```diff
@@ -550,8 +550,8 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	if sizeLimit == 0 {
 		sizeLimit = DefaultTxPoolConfig.TransactionSizeLimit
 	}
-	// Reject transactions over 64KB (or manually set limit) to prevent DOS attacks
-	if float64(tx.Size()) > float64(sizeLimit*1024) {
+	// Reject transactions over defined size to prevent DOS attacks
+	if uint64(tx.Size()) > sizeLimit*1024 {
```
Comment (baptiste-b-pegasys) [inline core/tx_pool.go]:
```suggestion
	if float64(tx.Size()) > float64(sizeLimit*1024) {
```

Comment (baptiste-b-pegasys) [inline core/tx_pool.go]:
```suggestion
	if uint64(tx.Size()) > sizeLimit*1024 {
```
```suggestion
	if float64(tx.Size()) > float64(sizeLimit*1024) {
```

File: core/tx_pool.go [hunk 4]
```diff
@@ -595,7 +595,6 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 		}
 	} else {
 		// Drop non-local transactions under our own minimal accepted gas price
-		local = local || pool.locals.contains(from) // account may be local even if the transaction arrived from the network
 		if !local && tx.GasPriceIntCmp(pool.gasPrice) < 0 {
```
Comment (baptiste-b-pegasys) [inline core/tx_pool.go]:
```suggestion
		local = local || pool.locals.contains(from) // account may be local even if the transaction arrived from the network
		if !local && tx.GasPriceIntCmp(pool.gasPrice) < 0 {
```

Comment (baptiste-b-pegasys) [inline core/tx_pool.go]:
```suggestion
		local = local || pool.locals.contains(from) // account may be local even if the transaction arrived from the network
		if !local && tx.GasPriceIntCmp(pool.gasPrice) < 0 {
```
```suggestion
		if !local && tx.GasPriceIntCmp(pool.gasPrice) < 0 {
```

File: eth/sync.go [hunk 5]
```diff
@@ -336,6 +336,10 @@ func (h *handler) doSync(op *chainSyncOp) error {
 		log.Info("Fast sync complete, auto disabling")
 		atomic.StoreUint32(&h.fastSync, 0)
 	}
+	/*if atomic.LoadUint32(&h.snapSync) == 1 {
+		log.Info("Snap sync complete, auto disabling")
+		atomic.StoreUint32(&h.snapSync, 0)
+	}*/
```
Comment (baptiste-b-pegasys) [inline eth/sync.go]:
```suggestion
	/* Disabled by Quorum
	if atomic.LoadUint32(&h.snapSync) == 1 {
		log.Info("Snap sync complete, auto disabling")
		atomic.StoreUint32(&h.snapSync, 0)
	}
	*/
```

Comment (baptiste-b-pegasys) [inline eth/sync.go]:
```suggestion
	/*
	if atomic.LoadUint32(&h.snapSync) == 1 {
		log.Info("Snap sync complete, auto disabling")
		atomic.StoreUint32(&h.snapSync, 0)
	}
	*/
```
```suggestion
	/*if atomic.LoadUint32(&h.snapSync) == 1 {
		log.Info("Snap sync complete, auto disabling")
		atomic.StoreUint32(&h.snapSync, 0)
	}*/
```

**Review by antonydenyer — APPROVED** [review]

File: eth/state_accessor.go [hunk 6]
```diff
@@ -99,7 +99,9 @@ func (eth *Ethereum) stateAtBlock(block *types.Block, reexec uint64) (statedb *s
 		if err != nil {
 			return nil, nil, nil, err
 		}
-		if err := statedb.Reset(root); err != nil {
+		err = statedb.Reset(root)
+		// statedb, err = state.New(root, database, nil) // Quorum
+		if err != nil {
```
Comment (achraf17) [inline eth/state_accessor.go:104]:
```suggestion
		if err := statedb.Reset(root); err != nil {
```

File: internal/ethapi/api.go [hunk 7]
```diff
@@ -2237,6 +2239,7 @@ func (s *PublicTransactionPoolAPI) SendRawTransaction(ctx context.Context, input
 func (s *PublicTransactionPoolAPI) SendRawPrivateTransaction(ctx context.Context, encodedTx hexutil.Bytes, args SendRawTxArgs) (common.Hash, error) {
 
 	tx := new(types.Transaction)
+	// if err := tx.UnmarshalBinary(encodedTx); err != nil { // Quorum
```
Comment (achraf17) [inline internal/ethapi/api.go:2242]: should we remove this?

File: node/node.go [hunk 8]
```diff
@@ -168,6 +168,16 @@ func New(conf *Config) (*Node, error) {
 	node.server.Config.DataDir = node.config.DataDir
 	// End Quorum
 
+	/* Quorum
+	// Check HTTP/WS prefixes are valid.
+	if err := validatePrefix("HTTP", conf.HTTPPathPrefix); err != nil {
```
Comment (achraf17) [inline node/node.go]: why are we removing this validation?

**Review by achraf17 — CHANGES_REQUESTED** [review]

File: internal/ethapi/api.go (hunk 7 above)
Comment (baptiste-b-pegasys) [reply internal/ethapi/api.go:2242]: it's a difference we have from the geth base.

File: node/node.go (hunk 8 above)
Comment (baptiste-b-pegasys) [reply node/node.go]: it's a difference we have from the geth base.

**Review by achraf17 — APPROVED** [review]
LGTM

**Pull request merged**